### PR TITLE
Reinstate UserEntered npc triggering

### PR DESCRIPTION
**Pull request by sicoderealm**
Allow NPCs to be informed of new users entering their zones to trigger things such as welcome messages. 

Moods are adjusted and dampened via the hourly `RecalcLightAndWeather` which in turn adjusts `AmbientLightChanged` thus `SetMood` (mood dampening)!

This will be a phase one that enables these welcome messages - after that we can add new messages and logic.

image

image

### CODE REVIEW

**Comment by Aesica** [comment]
Just my thoughts on the bullet points:

- Triggering:  The ideal *implementation* IMO would has some sort of short timer to avoid spamming.  You'd enter Familiars and Paddock would greet you, but if you left and came back in a few minutes later, you'd get nothing.  I'm not sure how ideal this would be in terms game data and code structure, though.  Each NPC would need a list of user/timer pairs which would be...unpleasant to work with.  Maybe best to just go with random *and* mood-based.  More likely to greet if the npc is in a good mood, for example.
- Mood adjustments:  As mentioned above, it could determine the rate of greetings.  It could also determine *how* the npc responds.  *Jonas greets you* vs *Jonas scowls at you.*  The level of detail could be per-npc, as not every npc needs to have detailed responses based on mood, and some don't care about outlaws/murderer status, factions, etc. 
- We'd definitely need to, especially if we want things like faction, outlaw, mood, etc to change responses.

Another thing to make note of is that some npc speech is in black text while other npc speech is in white text.  We should probably establish and adhere to a standard using just one or the other.

File: kod/object/active/holder/nomoveon/battler/monster/towns/barlqtwn/bqSmith.kod [hunk 1]
```diff
@@ -28,7 +28,7 @@ resources:
       "get the impression that he is uncomfortable with strangers, but his great size and clear "
       "devotion to Kraanan makes him afraid of no one."
 
-   BQsmith_entry_welcome = "Hello %s%s, and welcome."
+   BQsmith_entry_welcome = "~kFehr'loi Qan says, \"Hello %s%s, and welcome.\""
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/barlqtwn/bqSmith.kod:31]: I don't understand why in general these were all changed to include the name of the NPC here, duplicating it from elsewhere.  This makes the string more complicated and would also make translation more difficult in the future.  The previous way seems a lot better.

File: kod/object/active/holder/nomoveon/battler/monster/towns/hazartwn/hzinnk.kod [hunk 2]
```diff
@@ -46,5 +44,11 @@ properties:
 messages:
 
 
+   UserEntered(who = $)
+   {
+      % Marcus says hello as part of the initial welcoming text: guest1.kod
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/hazartwn/hzinnk.kod:49]: If someone changes guest1.kod, how will they know to update this file?

File: kod/object/active/holder/nomoveon/battler/monster/towns/marntwn/MrElder.kod [hunk 3]
```diff
@@ -87,9 +87,9 @@ messages:
       return;
    }
    
-   UserEntered(who=$)
+   UserEntered(who=$, what = $)
    {
-      Post(poOwner,@someoneSaid,#what=self,#string=MrEl_entry_welcome,
+      Post(who,@someoneSaid,#what=self,#string=MrEl_entry_welcome,
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/marntwn/MrElder.kod:92]: Why is this one posted to the user instead of the owner, unlike the others?

File: kod/object/active/holder/nomoveon/battler/monster/towns/marntwn/mrbart.kod [hunk 4]
```diff
@@ -31,7 +31,7 @@ resources:
 
    tova_unwanted_keep = "~k%s%s tells you, \"Well, how nice of you.  A gift.~n~k\""
    tova_unwanted_pay = "~k%s%s tells you, \"What an interesting item!~n~k\""
-   tova_unwanted_give = "~k%s%s tells you, \"Ha, ha, you've got to be kidding me, stranger! This is a tavern, not a gypsy�s tent!~n~k\""
+   tova_unwanted_give = "~k%s%s tells you, \"Ha, ha, you've got to be kidding me, stranger! This is a tavern, not a gypsy�s tent!~n~k\""
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/marntwn/mrbart.kod:34]: Not sure what the change is here.  Revert this file?

File: kod/object/active/holder/nomoveon/battler/monster/towns/marntwn/mrhealer.kod [hunk 5]
```diff
@@ -53,9 +53,8 @@ messages:
 
    UserEntered(who = $)
    {
-      Post(poOwner,@SomeoneSaid,#what=self,#string=MrHealer_entry_welcome,
-          #type = SAY_RESOURCE,
-          #parm1=Send(who,@GetDef),#parm2=Send(who,@GetName));
+
+      Post(poOwner,@SomeoneSaid,#what=self,#type=SAY_MESSAGE,#string=Mrhealer_entry_welcome);
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/marntwn/mrhealer.kod:57]: Does this work?  It looks like the substitution parameters have been removed.

File: kod/object/active/holder/nomoveon/battler/monster/towns/tostwn/TsSmith.kod [hunk 6]
```diff
@@ -93,9 +93,9 @@ messages:
 
    UserEntered(who = $)
    {
-      Post(poOwner,@SomeoneSaid,#what = self, #string=entry_welcome,
-         #type = SAY_RESOURCE, 
-         #parm1=Send(who,@GetDef),#parm2=Send(who,@GetName));
+
+      Post(poOwner,@SomeoneSaid,#what=self,#type=SAY_MESSAGE,#string=entry_welcome);
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/tostwn/TsSmith.kod:97]: Does this work?  It looks like the substitution parameters have been removed.

File: kod/object/active/holder/room.kod [hunk 7]
```diff
@@ -2419,6 +2421,35 @@ messages:
       }
 
       propagate;
+   }  
+   
+   UserEntered(who = $)
+   {
+      local i,each_obj,lastWelcomeTime;
+
+      if isClass(who,&Player)
```
Comment (akirmse) [inline kod/object/active/holder/room.kod:2430]: Is this IsClass check necessary?  It's called "UserEntered"

File: kod/object/active/holder/room.kod [hunk 8]
```diff
@@ -2419,6 +2421,35 @@ messages:
...
+      }
+
+      for i in plActive
+      {
+         each_obj = First(i);
+         % q: is there a better class than %Monster for all NPCs
+         if IsClass(each_obj, &Monster)
+         {
+            lastWelcomeTime = Send(each_obj, @GetLastWelcomeTime);
+            if GetTime() > (lastWelcomeTime + 30)
```
Comment (akirmse) [inline kod/object/active/holder/room.kod:2446]: I think this check belongs in Monster, not Room, since it's a behavior of the monster.

**Review by akirmse — COMMENTED** [review]
In general I think trying to restore this chatter is OK, but I don't like how the strings have been changed as a result.  Why was that necessary at all?

**Comment by akirmse** [comment]
Please revert all files where there's been no change, which I think is most of them.

Sometimes you've changed Send to Post, not sure why.

File: kod/object/active/holder/nomoveon/battler/monster/towns/kocatwn/kcvaultm.kod [hunk 9]
```diff
@@ -70,11 +70,11 @@ messages:
    UserEntered(who = $)
    {
       local iFlags;
-      if random(1,10)
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/kocatwn/kcvaultm.kod:73]: Why was this removed?

File: kod/object/active/holder/nomoveon/battler/monster.kod [hunk 10]
```diff
@@ -22,6 +22,8 @@ constants:
    %  learn to qualify for a new spell.
    MONSTER_LEARN_GRADIENT = 15
 
+   WELCOME_MESSAGE_INTERVAL = 30
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster.kod:25]: Add a comment including purpose and units.

File: kod/object/active/holder/nomoveon/battler/monster.kod [hunk 11]
```diff
@@ -5828,6 +5834,16 @@ messages:
       return;
    }
 
+   ReadyForWelcome()
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster.kod:5837]: Add a comment

File: kod/object/active/holder/room.kod [hunk 12]
```diff
@@ -1773,6 +1773,22 @@ messages:
...
 
+         for i in plActive
+         {
+            each_obj = First(i);
+            if IsClass(each_obj,&Monster)
+            {
+               if NOT Send(what,@CheckPlayerFlag,#flag=PFLAG_ANONYMOUS)
+                     AND NOT Send(what,@CheckPlayerFlag,#flag=PFLAG_INVISIBLE)
+               {
+                  if Send(each_obj, @ReadyForWelcome)
```
Comment (akirmse) [inline kod/object/active/holder/room.kod:1784]: How often does this result in multiple welcome messages upon one user entering the room?

File: kod/object/active/holder/room.kod [hunk 13]
```diff
@@ -1773,6 +1773,22 @@ messages:
                  #new_col=Nth(new_pos,2));
          }
 
+         for i in plActive
+         {
+            each_obj = First(i);
+            if IsClass(each_obj,&Monster)
+            {
+               if NOT Send(what,@CheckPlayerFlag,#flag=PFLAG_ANONYMOUS)
```
Comment (akirmse) [inline kod/object/active/holder/room.kod:1781]: Seems like these checks should be done outside the outer loop, not on each iteration.

File: kod/object/active/holder/nomoveon/battler/monster/towns/marntwn/MrSmith.kod [hunk 14]
```diff
@@ -92,7 +92,8 @@ messages:
 
    UserEntered(who = $)
    {
-      Post(poOwner,@SomeoneSaid,#what=self,#string=MrBs_entry_welcome,
+
+      Send(poOwner,@SomeoneSaid,#what=self,#string=MrBs_entry_welcome,
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/marntwn/MrSmith.kod:96]: Just verifying that you have tested every one of these instances where you changed Post to Send.  It's possible some were added intentionally to get the order of messages correct upon entering a room.  Please revert any you haven't tested.

File: kod/object/active/holder/nomoveon/battler/monster/towns/hazartwn/hzinnk.kod [hunk 15]
```diff
@@ -46,5 +44,12 @@ properties:
 messages:
 
 
+   UserEntered(who = $)
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/hazartwn/hzinnk.kod:47]: This function has no effect.  Remove?

File: kod/object/active/holder/nomoveon/battler/monster/towns/kocatwn/kcsmith.kod [hunk 16]
```diff
@@ -107,7 +107,7 @@ messages:
 
    UserEntered(who = $)
    {
-      Post(poOwner,@SomeoneSaid,#what = self, #string=kcSmith_entry_welcome,
+      Send(poOwner,@SomeoneSaid,#what = self, #string=kcSmith_entry_welcome,
```
Comment (akirmse) [inline kod/object/active/holder/nomoveon/battler/monster/towns/kocatwn/kcsmith.kod:110]: You still need to revert all of these changed files that are not related to this PR.
